### PR TITLE
UDP encryption fix

### PR DESCRIPTION
**Pull request by nico9889**
The main change I made is limiting the last encrypted block, appended to the encrypted string, to the actual remaining length.
I think previously some padding bytes were added and this caused a wrong checksum calculation, so the server were refusing most of the packets because of the wrong header tag (derived from the checksum).

To fix this I recompiled the Mumble server to add some useful log to find the error, but what led me to the solution was this very old implementation, found between Mumble issues, reported as working (at the time at least):
https://github.com/ianling/mumpy/blob/dev/mumpy/mumblecrypto.py

This is my first public contribution, I've put the same link in the commit description as a credit, I don't know if it's needed or if this is enough.

The last two changes are needed just to get everything at a minimum working stage:
* I added a default length to the decrypt method, which is the packet length minus the header (1 byte of the iv, 3 bytes of tag)~~, sometimes along the code it wasn't specified as an argument and that made the library crash~~ Edit: manual parameter specification has been completely removed as suggested by @dkirker since it was error prone;
* I renamed a method call in soundoutput, since send_udp wasn't an existing method.

After the changes, all the tests in test_crypto.py still passes and Mumble reports the UDP protocol as working (the number of the packet sent is increasing while reproducing music).

### CODE REVIEW

File: pymumble_py3/crypto.py
Comment (dkirker) [inline pymumble_py3/crypto.py]: I found out that it was best to just remove the `len_plain` as a parameter to `decrypt` and just use this line. The problem that I hit was that if `len_plain` was passed in, it contained the length including the header and that caused issues with decrypting later on. The call to `decrypt` basically just passes `len_plain` as `len(source)` anyway.

**Comment by dkirker** [comment]
On top of ditching `len_plain` as a parameter to `ocb.decrypt` I found that I had to remove a lot of unnecessary code in `read_udp_packet` and a rouge `recvfrom` in `loop`. I haven't actually tested to confirm that received audio is actually valid sounding audio, but the logging shows properly incrementing sequence numbers from the received data.

**Comment by nico9889** [comment]
The only purpose of this pull request is to fix only the encryption function. I may be wrong, but the whole UDP related code to me it looks like a "quick patch" with the only purpose of test the encryption rather than an actual final implementation.

In the original code, the `decrypt` function was used both with and without `len_plain`, you can see that looking at lines [226](https://github.com/azlux/pymumble/blob/9f2fa373ceb67777c7b1d1f181bd0bf5ac5605f5/pymumble_py3/mumble.py#L226) and [336](https://github.com/azlux/pymumble/blob/9f2fa373ceb67777c7b1d1f181bd0bf5ac5605f5/pymumble_py3/mumble.py#L336) in mumble.py, but this was dead code since UDP wouldn't initialize because of the wrong ping encryption.

That's the reason why I "patched" the decrypt function making it work both with and without `len_plain` specified, letting the final choice the maintainers of the repository. 

For the same reason, I wouldn't change `read_udp_packet` in this pull request, I would rather wait for @azlux (or anyone other authorized) to merge this fix so a proper implementation can be done later.

**Comment by dkirker** [comment]
What I am suggesting is to simplify the contract for `CryptStateOCB2.decrypt` by removing the `len_plain` parameter and remove the `if not len_plain:` line so that only `len_plain = len(source) - 4  # The entire packet length, minus the header` is left. The issue with having the `len_plain` parameter in place is that the call on line 336 in mumble.py is resulting in the following exception:

> Traceback (most recent call last):
>  File "/usr/lib/python3.8/threading.py", line 932, in _bootstrap_inner
>    self.run()
>  File "/usr/local/lib/python3.8/dist-packages/pymumble-1.6.1-py3.8.egg/pymumble_py3/mumble.py", line 139, in run
>  File "/usr/local/lib/python3.8/dist-packages/pymumble-1.6.1-py3.8.egg/pymumble_py3/mumble.py", line 283, in loop
>  File "/usr/local/lib/python3.8/dist-packages/pymumble-1.6.1-py3.8.egg/pymumble_py3/mumble.py", line 342, in read_udp_packet
>  File "/usr/local/lib/python3.8/dist-packages/pymumble-1.6.1-py3.8.egg/pymumble_py3/crypto.py", line 217, in decrypt
>  File "/usr/local/lib/python3.8/dist-packages/pymumble-1.6.1-py3.8.egg/pymumble_py3/crypto.py", line 356, in ocb_decrypt
>  File "/usr/local/lib/python3.8/dist-packages/pycryptodome-3.17-py3.8-linux-x86_64.egg/Crypto/Cipher/_mode_ecb.py", line 141, in encrypt
>    raise ValueError("Data must be aligned to block boundary in ECB mode")
>ValueError: Data must be aligned to block boundary in ECB mode

(I have some other code in my mumble.py, so the line numbers for those above differ; but my crypto.py is basically the same.)

The alternative is to have any call to `decrypt` that *does* include the `len_plain` parameter make sure that it is `len(source)-4`, like `decrypt(source, len(source)-4)`. But I think just removing the `len_plain` parameter for `decrypt` (so you just call `decrypt(source)`) and just using your `len_plain = len(source) - 4  # The entire packet length, minus the header` line is the easiest and cleanest. Hopefully I've better described what I am suggesting.

I guess my reason for the suggestion is that I can't really think of a reason where having the `len_plain` as a passable parameter to `CryptStateOCB2.decrypt` makes sense.

With that change, your changes work perfect!

**Comment by nico9889** [comment]
> What I am suggesting is to simplify the contract for `CryptStateOCB2.decrypt` by removing the `len_plain` parameter and remove the `if not len_plain:` line so that only `len_plain = len(source) - 4 # The entire packet length, minus the header` is left. The issue with having the `len_plain` parameter in place is that the call on line 336 in mumble.py is resulting in the following exception:
> 
> > Traceback (most recent call last):
> > File "/usr/lib/python3.8/threading.py", line 932, in _bootstrap_inner
> > self.run()
> > File "/usr/local/lib/python3.8/dist-packages/pymumble-1.6.1-py3.8.egg/pymumble_py3/mumble.py", line 139, in run
> > File "/usr/local/lib/python3.8/dist-packages/pymumble-1.6.1-py3.8.egg/pymumble_py3/mumble.py", line 283, in loop
> > File "/usr/local/lib/python3.8/dist-packages/pymumble-1.6.1-py3.8.egg/pymumble_py3/mumble.py", line 342, in read_udp_packet
> > File "/usr/local/lib/python3.8/dist-packages/pymumble-1.6.1-py3.8.egg/pymumble_py3/crypto.py", line 217, in decrypt
> > File "/usr/local/lib/python3.8/dist-packages/pymumble-1.6.1-py3.8.egg/pymumble_py3/crypto.py", line 356, in ocb_decrypt
> > File "/usr/local/lib/python3.8/dist-packages/pycryptodome-3.17-py3.8-linux-x86_64.egg/Crypto/Cipher/_mode_ecb.py", line 141, in encrypt
> > raise ValueError("Data must be aligned to block boundary in ECB mode")
> > ValueError: Data must be aligned to block boundary in ECB mode

Thank you! I completely missed that! It has been a while since I opened this pull request and I thought that I tested everything at the time.

> (I have some other code in my mumble.py, so the line numbers for those above differ; but my crypto.py is basically the same.)
> 
> The alternative is to have any call to `decrypt` that _does_ include the `len_plain` parameter make sure that it is `len(source)-4`, like `decrypt(source, len(source)-4)`. But I think just removing the `len_plain` parameter for `decrypt` (so you just call `decrypt(source)`) and just using your `len_plain = len(source) - 4 # The entire packet length, minus the header` line is the easiest and cleanest. Hopefully I've better described what I am suggesting.
> 
> I guess my reason for the suggestion is that I can't really think of a reason where having the `len_plain` as a passable parameter to `CryptStateOCB2.decrypt` makes sense.

I agree, I was just trying to do less modification as possible to the original code focusing only on the crypto, but at this point I can't think any other reason to avoid the change. I'll fix it in some minutes.

File: pymumble_py3/crypto.py
Comment (nico9889) [reply pymumble_py3/crypto.py]: This should be solved now

**Comment by dkirker** [comment]
@nico9889 No worries! I might not have been completely clear in my original comment inline in the code.

@azlux I can confirm that this code works!

When this merges I'll create another PR to clean up the `read_udp_packet` function.

**Review by dkirker — APPROVED** [review]

**Pull request merged**
